compiler/rustc_middle/src/hir/place.rs

```diff
@@ -113,3 +113,50 @@ impl<'tcx> Place<'tcx> {
         if projection_index == 0 { self.base_ty } else { self.projections[projection_index - 1].ty }
     }
 }
+
+pub fn place_to_string_for_capture(place: &Place<'tcx>, base_name: String) -> String {
+    let mut curr_string = format!("{}", base_name);
+    for (i, proj) in place.projections.iter().enumerate() {
+        if ProjectionKind::Deref == proj.kind {
+            curr_string = format!("*{}", curr_string);
+        } else {
+            match place.ty_before_projection(i).kind() {
+                ty::Adt(def, ..) if def.is_enum() => {
+                    // we might be projecting *to* a variant, or to a field *in* a variant.
+                    match proj.kind {
+                        ProjectionKind::Field(_idx, variant) => {
+                            curr_string = format!(
+                                "{}.{}",
+                                curr_string,
+                                def.variants[variant].fields[0].ident.name.as_str()
+                            );
+                        }
+                        _ => {
+                            bug!("Could not handle ProjectionKind {:?}", proj.kind)
+                        }
+                    }
+                }
+                ty::Adt(def, ..) => {
+                    curr_string = format!(
+                        "{}.{}",
+                        curr_string,
+                        def.non_enum_variant().fields[0].ident.name.as_str()
```

Comment:
we might not have an indent in case of tuple struct pattern, can u add a test case something like

```rust
#[!feature(capture_disjoint_fields)]
struct S(String);

fn expect_fn<F: Fn()>(_f: F) {}

fn main() {
    let s = S(format!("s");
    let c = || {
        let s = s.0;
    };
    expect_fn(c);
}
```

Reply:
Done, also added one for using structs too